File: ds_algo/grind75/1_TwoSum.py
### GRIND 75 SECOND PASS ###
"""
approach:
    brute force:
        try each element added with each other element
            possible to index i, j = i+1 to only get below the diagonal on the matrix and be a bit more efficient
        return when we see a solution cause we know there's only one solution
    
    hash map:
        loop through the list, store the additive complement of each element in a dict (complement is the key, index is the val)
        check if the current nums val is already recorded as a complement
            if so, return the current index and the index indicated by the complement
    
    sorted array:
        sort the array & keep track of their original indices (e.g. convert to a list of tuples)
        start from the large side of the array 
        take this number as the anchor, call it "larger"
        (you might be tempted to take "larger" as the next smallest value below anchor, but this doesn't work with negatives in list)
        outer loop: 
            inner loop:
                starting from the smallest number and continuing to just below larger, add it to the larger and compare to target
                return if the sum == target
                break if sum > target
            once the inner loop is finished, move anchor to the next smallest number
            
    two pointers: (improvement on sorted array)
        sort the array and keep track of original indices
        start pointers at the righthand side and lefthand side 
        loop through the list
            check if the left and right pointers sum to target (if so, return True)
            if not, compare sum to target
            if sum is greater, lower the righthand (larger-value) pointer to the next lowest value
            if sum is smaller, raise the lefthand (lower-value) pointer to the next highest value
"""

# two pointers (better sorted array)
# T: O(n)
# S: O(1)
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        # track original indices, sort as tuple, cast back to list
        length  = len(nums)
        indices = list(range(length))
        nums    = list(zip(nums, indices))
        nums.sort()
        right   = length - 1 # righthand pointer starts at the last element
        left    = 0 # lefthand pointer starts at first element
        while True: # bad form but problem description guarantees a solution
            # add left and right pointers
            res = nums[left][0] + nums[right][0]
            if res == target:
                lidx, ridx = nums[left][1], nums[right][1] # get the original indices for these numbers
                return [lidx, ridx]
            elif res < target:
                left += 1 # raise left if sum is lower than target
            else:
                right -= 1 # lower right if sum is greater than target
                    
            

# sorted array 
# T: O(n^2) at the worst, but likely much better
# S: O(n)
class __Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        # track original indices, sort as tuple, cast back to list
        length = len(nums)
        indices = list(range(length))
        nums = list(zip(nums, indices))
        nums.sort()
        # righthand bound is at the righthand side of the list
        larger = length - 1
        # find suitable pair
        while larger > 0:
            smaller = 0
            while smaller < larger:
                res = nums[larger][0] + nums[smaller][0]
                if res == target:
                    # translate to the original indices
                    smalleridx, largeridx = nums[smaller][1], nums[larger][1]
                    return [smalleridx, largeridx]
                smaller += 1
            larger -= 1
            

# hash map
# T: O(n) (as long as the hash lookup is O(1))
# S: O(n) (space for the hash map)
class __Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        hashmap = {} # python dict
        for i in range(len(nums)):
            current = nums[i]
            if current in hashmap:
                return [hashmap[current], i]
            complement = target - current
            hashmap[complement] = i
                    

# brute force
# T: O(n^2)
# S: O(1)
class __Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        for i in range(1, len(nums)):
            for j in range(0, i):
                r, c = nums[i], nums[j]
                # print(c, ", ", r)
                res = r + c
                if res == target:
                    return [i,j]

File: ds_algo/grind75/1_TwoSum.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */

/*
    approach:
        brute force: O(n^)
            two loops nested, check the sum of each against target
            skip cases where the loop indices are the same (same elem can't be used with itself)
            return early if sum is found
            return nothing otherwise (we assume a solution exists)
        
        hash map: O(n)
            declare a hashmap (dict or obj), which will contain key:value as sumcomplement:index
            loop through the array (only once)
                on each iteration, calculate complement = target - nums[i]
                check if the hashmap already contains the complement
                    if it does, then return the index(dict value) at the complement, and i
                if it does not, then add nums[i] to the hashmap (NOT add the complement)
        
        two pointer: O(n*log(n))
            the O-analysis is valid if we need to return numbers instead of indices
                otherwise, we'll need to search the original array for the indices -> add O(2n)
            create a copy of the array and sort the new array
            create two pointers ("left" and "right")
*/

// HASH MAP APPROACH
var twoSum = function (nums, target) {
  let hash = {};
  let complement; // will be target - nums[i]
  // loop over nums, check if complement is in hash, add nums[i] if not, return if found
  for (let i = 0; i < nums.length; i++) {
    complement = target - nums[i];
    if (hash.hasOwnProperty(complement)) { // if our complement exists, then we have a match
      // hash[complement] contains the index of the complement in nums
      // i is the index of the current element
      return [hash[complement], i];
    } else {
      // if we didn't find a match, store the current element and index in the map
      // this ensures that we will get a match in the hashmap for the current elem's complement when we see it later
      hash[nums[i]] = i;
    }
  }
};

// TWO POINTER APPROACH
// var twoSum = function(nums, target) {
//     // copy the array so we can preserve the original array order
//     sorted = [...nums];
//     sorted.sort((a, b) => a - b); // javascript makes me lose faith in humanity sometimes
//     // set up two "pointers" (in js, these will just be vars representing indices)
//     // "left" will point to the first index, "right" will point to the last index
//     let left = 0;
//     let right = nums.length - 1;
//     // loop while left < right (if they converge and we have not found, then there is no match)
//     do {
//         sum = sorted[left] + sorted[right];
//         // if they sum to target, we have a match so return the indices
//         // if they sum to less than target, then we need larger numbers 
//         // therefore increment left because we already checked the larger numbers right of right
//         // if they sum to greater than target, we need smaller numbers
//         // therefore decrement right because we already checked the smaller numbers left of left
//         // if this doesn't make sense, walk through the first few cases
//         // it should make sense then how we won't "miss" any potential combos
//         if(sum === target) {
//             // since we sorted the array, our indices don't map to the original array
//             // need to find the indices of the original array
//             // this breaks the O(n*logn) but would work in a similar problem where we only return nums, not indices
//             let newLeft = nums.findIndex(elem => elem === sorted[left]);
//             let newRight = nums.findIndex(elem => elem === sorted[right]);
//             // since we might have the same element twice and find will give us the first occurrence, we need to check if left and right are equivalent
//             // if they are, we can find sorted[right] in the subarray from newLeft + 1 onwards
//             if (newLeft === newRight) {
//                 let offset = newLeft + 1;
//                 newRight = offset + nums.slice(offset).findIndex(elem => elem === sorted[right]); 
//             }
//             return [newLeft, newRight];
//         } else if (sum < target) {
//             left++;
//         } else if (sum > target) {
//             right--;
//         }
//     } while (left < right);
// };

// BRUTE FORCE APPROACH
// var twoSum = function(nums, target) {
//     // nested loops
//     for(let i = 0; i < nums.length; i++) {
//         // start the inner loop just ahead of the outer to stay "below the diagonal" (no repeats)
//         for(let k = i+1; k < nums.length; k++) {
//             if(nums[i] + nums[k] === target) {
//                 return [i, k];
//             }
//         }
//     }
// };